### DISCUSSION TITLE
LibreVNA External Power and connection

### DISCUSSION THREAD

**Discussion by Michaelli333** [jankae/LibreVNA#229] [Q&A]
Hi Jan
I have two questions. The first one is about librevna external power. I found that the external power supply interface controls the power supply mode by detecting whether the plug is inserted. I want to design an external switch button to switch between USB power supply and external power supply. How should I implement it?
The second  one is about the connection with the VNA. Sometimes, when I open the gui and plug the VNA into the computer,  Connected to <****>   is displayed in the lower right corner, but the firmware number is not displayed on the right and   No device information available yet   is displayed. How should I solve it?
Best regards
Li

**Comment by jankae** [comment]
Hi Li,

> I want to design an external switch button to switch between USB power supply and external power supply. How should I implement it?

I think this is only possible by modifying the LibreVNA circuit. If you have the necessary skills, it should be fairly easy to figure out in which way you need to change the schematic but actually doing the modification can get a bit tricky (due to the internal layers in the PCB).
An idea to get you started: cut the trace from the DC connector and connect it to a switch like this:
![image](https://github.com/jankae/LibreVNA/assets/16818866/938cf6a7-e314-4170-b75a-34e7d24b2247)
This should at least give you the ability to switch on/off the supply from the USB connector. The problem is that the 5V from USB will appear at the DC connector when it is switched on. This could cause issues if you have something connected to the DC jack.

> Sometimes, when I open the gui and plug the VNA into the computer, Connected to <****> is displayed in the lower right corner, but the firmware number is not displayed on the right and No device information available yet is displayed. How should I solve it?

This means that the GUI hasn't received an answer to the RequestDeviceInfo packet (or hasn't handled that packet correctly). This is most likely a bug. You can:

- disconnect and reconnect to the device which should solve this
- use a newer version of the firmware/GUI (I think I remember the problem from some time ago but haven't seen it for quite some time)
- if you are already using 1.5.0-alpha.2 and still experience the problem, please try to figure out what you can do to trigger the problem and create an issue for that. I will be happy to fix that if I can replicate it

**Reply by jankae** [reply]
> but when I change a computer and connect in the same way, this problem occurs.

Please try this again with 1.5.0. I think someone had a similar issue before and the fix has been implemented in the new version.

> Could it be that using a new thread to call the GUI caused this problem?

I can't think of anything thread-related that could cause the problem here. The GUI is an external program anyway and as long as it is running, you should be able to perform all API calls.

> Because the connection can only be established when the GUI is open. Is there any other way for me to establish the connection in python to avoid this problem？

I am not quite following here. There are several connections and I am not sure which you mean:
- physical USB connection from the LibreVNA to the computer
- libusb connection from the PC to the endpoints of the USB device
- logical connection inside of the GUI from the libusb device handle to the abstract DeviceDriver class
- telnet connection from your python script to the SCPI server in the GUI

With the exception of the first one, none of these can be established without running the GUI.

**Reply by Michaelli333** [reply]
Thanks a lot. I will have a try.
